### PR TITLE
Add the basics task solution

### PR DESCRIPTION
**Pull request by TheVitik**

### CODE REVIEW

File: src/basics/Basics.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+namespace basics;
+
+use InvalidArgumentException;
+
+class Basics implements BasicsInterface
+{
+    private BasicsValidator $validator;
+
+    public function __construct(BasicsValidator $validator)
```
Comment (oleksandr-chornovol) [inline src/basics/Basics.php]: Please change it to `public function __construct(private BasicsValidator $validator)`

File: src/basics/Basics.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * The $minute variable contains a number from 0 to 59 (i.e. 10 or 25 or 60 etc).
+     * Determine in which quarter of an hour the number falls.
+     * Return one of the values: "first", "second", "third" and "fourth".
+     * Throw InvalidArgumentException if $minute is negative or greater than 60.
+     * (Implement this functionality in isMinutesException method from BasicsValidator Class and use it here)
+     * @see https://www.php.net/manual/en/class.invalidargumentexception.php
+     *
+     * Make sure the next PHPDoc instructions will be added or use typehint:
+     * @param int $minute
```
Comment (oleksandr-chornovol) [inline src/basics/Basics.php]: Please don't use annotations for `int`, `bool` and `string`

File: src/basics/Basics.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+     * @return string
+     * @throws InvalidArgumentException
+     */
+    public function getMinuteQuarter(int $minute): string
+    {
+        $this->validator->isMinutesException($minute);
+
+        if($minute>45 || $minute==0){
```
Comment (oleksandr-chornovol) [inline src/basics/Basics.php]: Please use match instead of `if else`

File: src/basics/Basics.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            return "fourth";
```
Comment (oleksandr-chornovol) [inline src/basics/Basics.php]: Please use single quotes for such cases

Comment (oleksandr-chornovol) [inline src/basics/Basics.php]: Please use constants for such strings

File: src/basics/Basics.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+     * Make sure the next PHPDoc instructions will be added:
+     * @param int $year
+     * @return boolean
+     * @throws InvalidArgumentException
+     */
+    public function isLeapYear(int $year): bool
+    {
+        $this->validator->isYearException($year);
+
+        if($year%4==0){
```
Comment (oleksandr-chornovol) [inline src/basics/Basics.php]: Please make it a single line.
Like that `return ($year % 4 == 0 && $year % 100 != 0) || ($year % 400 == 0);`

File: src/basics/Basics.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+     * Make sure the next PHPDoc instructions will be added:
+     * @param string $input
+     * @return boolean
+     * @throws InvalidArgumentException
+     */
+    public function isSumEqual(string $input): bool
+    {
+        $this->validator->isValidStringException($input);
+
+        $length = strlen($input);
```
Comment (oleksandr-chornovol) [inline src/basics/Basics.php]: The same here
`return (array_sum(str_split(substr($input, 0, 3))) === array_sum(str_split(substr($input, 3))));`

File: src/basics/Basics.php [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+            if($i+1<=$length/2){
+                $sumOfFirst+=(int) $input[$i];
+            }
+            else{
+                $sumOfSecond+=(int) $input[$i];
+            }
+        }
+        return $sumOfFirst == $sumOfSecond;
+    }
+}
```
Comment (oleksandr-chornovol) [inline src/basics/Basics.php]: Please add an empty line in the end of the file

File: src/basics/BasicsValidator.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * Implement the check functionality described in the method getMinuteQuarter (BasicsInterface Class) which throws Exception.
+     *
+     * Make sure the next PHPDoc instructions will be added:
+     * @param int $minute
+     * @throws InvalidArgumentException
+     */
+    public function isMinutesException(int $minute): void
+    {
+        if($minute<0 || $minute>60){
```
Comment (oleksandr-chornovol) [inline src/basics/BasicsValidator.php]: Please refactor it according to the [PSR-12](https://www.php-fig.org/psr/psr-12/)

File: src/basics/BasicsValidator.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+            throw new InvalidArgumentException("The minute variable contains a number from 0 to 59");
```
Comment (oleksandr-chornovol) [inline src/basics/BasicsValidator.php]: Please use single quotes is such cases

**Review by oleksandr-chornovol — CHANGES_REQUESTED** [review]
Please fix all comments

File: src/basics/Basics.php [hunk 10]
```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace basics;
+
+use InvalidArgumentException;
+
+class Basics implements BasicsInterface
+{
+
```
Comment (oleksandr-chornovol) [inline src/basics/Basics.php]: Please remove the empty line

File: src/basics/Basics.php [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    private const FIRST = 'first';
+    private const SECOND = 'second';
+    private const THIRD = 'third';
+    private const FOURTH = 'fourth';
+
```
Comment (oleksandr-chornovol) [inline src/basics/Basics.php:13]: Please remove the empty line

File: src/basics/Basics.php [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    public function __construct(private BasicsValidator $validator){}
+
+    /**
+     * @throws InvalidArgumentException
+     */
+    public function getMinuteQuarter(int $minute): string
+    {
+        $this->validator->isMinutesException($minute);
+
+        return match(true){
```
Comment (oleksandr-chornovol) [inline src/basics/Basics.php]: Please refactor it according to the PSR-12

File: src/basics/Basics.php [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+            $minute > 45 || $minute == 0 => self::FOURTH,
```
Comment (oleksandr-chornovol) [inline src/basics/Basics.php:24]: Are you sure that you should return 'fourth' for 0?

File: src/basics/BasicsValidator.php [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+class BasicsValidator implements BasicsValidatorInterface
+{
+
+    /**
+     * @throws InvalidArgumentException
+     */
+    public function isMinutesException(int $minute): void
+    {
+        if($minute < 0 || $minute > 60){
```
Comment (oleksandr-chornovol) [inline src/basics/BasicsValidator.php]: Please refactor it according to the PSR-12

File: src/basics/BasicsValidator.php [hunk 15]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace basics;
+
+use InvalidArgumentException;
+
+class BasicsValidator implements BasicsValidatorInterface
+{
+
```
Comment (oleksandr-chornovol) [inline src/basics/BasicsValidator.php]: Please remove the empty line

**Review by oleksandr-chornovol — CHANGES_REQUESTED** [review]
Please fix all comments

File: src/basics/Basics.php (hunk 13 above)
Comment (TheVitik) [reply src/basics/Basics.php:24]: Yes, according to the tests, 0 = 60 = fourth

**Review by oleksandr-chornovol — APPROVED** [review]
Good job!

**Pull request merged**